r1cs/gadgets/std/src/fields/nonnative/reduce.rs

```diff
@@ -252,25 +421,95 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField> Reducer<SimulationF, Cons
             ));
         }
 
-        // This part we mostly use the techniques in bellman-bignat
-        // The following code is adapted from https://github.com/alex-ozdemir/bellman-bignat/blob/master/src/mp/bignat.rs#L567
+        // The equality of two `A`-ary representations `[L[0],L[1],..]` and `[R[0],R[1],..]` with 
+        // oversized limbs is proven by enforcing their *shifted* differences 
+        // ``
+        //      [L[0] + (- R[0] + shift_constant) , 
+        //          L[1] + (- R[1] + shift_constant), 
+        //              ..],
+        // `` 
+        // to represent 
+        // ``
+        //      Sum_{i>=0} shift_constant * A^i.
+        // ``
+        // The constant `shift_constant = 2^bits_per_group  - 1` is to circumvent underflows in a
+        // length-preserving manner. With this choice  
+        // ``
+        //     0 <=  shift_constant - R[i] < 2^bits_per_group,
+        // `` 
+        // and hence
+        // ``
+        //      L[i] + (shift_constant - R[i]) < 2^{bits_per_group + 1},
+        // ``
+        // Since the length of the carries are throughout `<= bits_per_limb + surfeit`, 
+        // we also have `carry[i-1] + L[i] < 2^bits_per_group`, and the overall sum 
+        // ``
+        //      carry[i-1] + L[i] + (shift_constant - R[i]) 
+        // `` 
+        // is still of length `<= bits_per_group + 1`.
+        
+        // Why carries don't effect the length bound: 
+        // Assuming `len(carry) <= bits_per_limb + surfeit`, which is shown in detail below, 
+        // `` 
+        //   carry + limb[0] + limb[1] * A < 2 * A * 2^limb_size
+        // ``
+        // and therefore 
+        // ``
+        //   carry + limb[0] + limb[1] * A + limb[2] * A^2 < 2 * A^2 * 2^limb_size  
+        //   ...
+        //   carry + limb[0] + limb[1] * A + ... + limb[S-1] * A^{S-1} <
+        //                       < 2 * A^{S-1} * 2^limb_size.
+        // ``
```

Comment:
I do not understand this inequality: it should be  < carry + (1 + A) * 2^limb_size.  I accept anyway that carry < 2^(limb_size - shift_per_limb + 1), so the inequality holds a fortiori.